File: go/44.go
/*

https://leetcode.com/problems/wildcard-matching/

Given an input string (s) and a pattern (p), implement wildcard pattern matching with support for '?' and '*'.

'?' Matches any single character.
'*' Matches any sequence of characters (including the empty sequence).
The matching should cover the entire input string (not partial).

Note:

s could be empty and contains only lowercase letters a-z.
p could be empty and contains only lowercase letters a-z, and characters like ? or *.
Example 1:

Input:
s = "aa"
p = "a"
Output: false
Explanation: "a" does not match the entire string "aa".
Example 2:

Input:
s = "aa"
p = "*"
Output: true
Explanation: '*' matches any sequence.
Example 3:

Input:
s = "cb"
p = "?a"
Output: false
Explanation: '?' matches 'c', but the second letter is 'a', which does not match 'b'.
Example 4:

Input:
s = "adceb"
p = "*a*b"
Output: true
Explanation: The first '*' matches the empty sequence, while the second '*' matches the substring "dce".
Example 5:

Input:
s = "acdcb"
p = "a*c?b"
Output: false

*/

package main


import "fmt"

// naive

/*
func isMatch(s string, p string) bool {
    var f func(s string, p string) bool
    f = func(s string, p string) bool {
        if len(p) == 0 {
            return len(s) == 0
        }
        if len(s) == 0 {
            if rune(p[0]) == '*' {
                return f(s, p[1:])
            }
            return false
        }
        if rune(p[0]) == '*' {
            return f(s[1:], p[1:]) || f(s, p[1:]) || f(s[1:], p)
        }
        if s[0] == p[0] || rune(p[0]) == '?' {
            return f(s[1:], p[1:])
        }
        return false
    }
    return f(s, p)
}
*/

// memo

/*
func isMatch(s string, p string) bool {
    dict := make(map[[2]int]bool)
    var f func(s string, p string) bool
    f = func(s string, p string) bool {
        m, n := len(s), len(p)
        key := [2]int{m, n}
        if v, ok := dict[key]; ok {
            return v
        }
        if n == 0 {
            return m == 0
        }
        if m == 0 {
            if rune(p[0]) == '*' {
                r := f(s, p[1:])
                dict[key] = r
                return r
            }
            dict[key] = false
            return false
        }
        if rune(p[0]) == '*' {
            r := f(s, p[1:]) || f(s[1:], p) || f(s[1:], p[1:])
            dict[key] = r
            return r
        }
        if s[0] == p[0] || rune(p[0]) == '?' {
            r := f(s[1:], p[1:])
            dict[key] = r
            return r
        }
        dict[key] = false
        return false
    }
    return f(s, p)
}
*/


func isMatch(s string, p string) bool {
    m, n := len(s), len(p)
    dp := make([][]bool, m+1)
    for i := 0; i < m + 1; i++ {
        dp[i] = make([]bool, n + 1)
    }
    dp[m][n] = true
    for i := m; i >= 0; i-- {
        for j := n - 1; j >= 0; j-- {
            if rune(p[j]) == '*' {
                if i < m {
                    dp[i][j] = dp[i+1][j] || dp[i][j+1] || dp[i+1][j+1]
                } else {
                    dp[i][j] = dp[i][j+1]
                }
            } else if i < m && (s[i] == p[j] || rune(p[j]) == '?') {
                dp[i][j] = dp[i+1][j+1]
            }
        }
    }
    return dp[0][0]
}


// 凭借印象过了


func main() {
    s := "acdcb"
    p := "a*c?b"
    fmt.Println(isMatch(s, p))

    s = "aa"
    p = "*"
    fmt.Println(isMatch(s, p))

    s = "cb"
    p = "?a"
    fmt.Println(isMatch(s, p))

    s = "aa"
    p = "a"
    fmt.Println(isMatch(s, p))

    s = "adceb"
    p = "*a*b"
    fmt.Println(isMatch(s, p))

    s = "ab"
    p = "?*"
    fmt.Println(isMatch(s, p))

    s = "aaabbbaabaaaaababaabaaabbabbbbbbbbaabababbabbbaaaaba"
    p = "a*******b"
    fmt.Println(isMatch(s, p))

}

File: python/44.py
"""
https://leetcode.com/problems/wildcard-matching/

Given an input string (s) and a pattern (p), implement wildcard pattern matching with support for '?' and '*'.

'?' Matches any single character.
'*' Matches any sequence of characters (including the empty sequence).
The matching should cover the entire input string (not partial).

Note:

s could be empty and contains only lowercase letters a-z.
p could be empty and contains only lowercase letters a-z, and characters like ? or *.
Example 1:

Input:
s = "aa"
p = "a"
Output: false
Explanation: "a" does not match the entire string "aa".
Example 2:

Input:
s = "aa"
p = "*"
Output: true
Explanation: '*' matches any sequence.
Example 3:

Input:
s = "cb"
p = "?a"
Output: false
Explanation: '?' matches 'c', but the second letter is 'a', which does not match 'b'.
Example 4:

Input:
s = "adceb"
p = "*a*b"
Output: true
Explanation: The first '*' matches the empty sequence, while the second '*' matches the substring "dce".
Example 5:

Input:
s = "acdcb"
p = "a*c?b"
Output: false
"""


class Solution1:
    def isMatch(self, s: str, p: str) -> bool:
        from functools import lru_cache

        @lru_cache
        def f(s, p):
            if not p:
                return not s
            if not s:
                return p[0] == '*' and f(s, p[1:])
            if s[0] == p[0] or p[0] == '?':
                return f(s[1:], p[1:])
            if p[0] == '*':
                return f(s[1:], p[1:]) or f(s, p[1:]) or f(s[1:], p)
            return False

        return f(s, p)


class Solution2:
    def isMatch(self, s: str, p: str) -> bool:
        m, n = len(s), len(p)
        memo = {}

        def f(i, j):
            key = (i, j)
            if key in memo:
                return memo[key]
            if j == n:
                r = i == m
                memo[key] = r
                return r
            if i == m:
                r = p[j] == '*' and f(i, j + 1)
                memo[key] = r
                return r
            if s[i] == p[j] or p[j] == '?':
                r = f(i + 1, j + 1)
                memo[key] = r
                return r
            if p[j] == '*':
                r = f(i + 1, j + 1) or f(i + 1, j) or f(i, j + 1)
                memo[key] = r
                return r
            return False

        return f(0, 0)


# 为什么这个递归的方式不行呢?
# 为什么加上缓存之后又可以了?
# 说明还是没有彻底搞懂啊?

class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        m, n = len(s), len(p)
        dp = [[False] * (n + 1) for _ in range(m + 1)]
        dp[-1][-1] = True
        for i in range(m, -1, -1):
            for j in range(n - 1, -1, -1):
                if p[j] == '*':
                    dp[i][j] = i < m and (dp[i + 1][j + 1] or dp[i + 1][j]) or dp[i][j + 1]
                elif i < m and (s[i] == p[j] or p[j] == '?'):
                    dp[i][j] = dp[i + 1][j + 1]

        return dp[0][0]




























from functools import lru_cache

class Solution:
    def isMatch(self, s: str, p: str) -> bool:

        @lru_cache
        def f(s, p):
            if not p:
                return not s
            if not s:
                return f(s, p[1:]) if p[0] == '*' else False
            if p[0] == '*':
                return f(s[1:], p) or f(s[1:], p[1:]) or f(s, p[1:])
            if s[0] == p[0] or p[0] == '?':
                return f(s[1:], p[1:])

            return False

        return f(s, p)



# 不用字符串 改为指针 速度会快一点
from functools import lru_cache

class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        m, n = len(s), len(p)

        @lru_cache
        def f(i, j):
            if j == n:
                return i == m
            if i == m:
                return f(i, j+1) if p[j] == '*' else False
            if p[j] == '*':
                return f(i+1, j) or f(i+1, j+1) or f(i, j+1)
            if s[i] == p[j] or p[j] == '?':
                return f(i+1, j+1)

            return False

        return f(0, 0)


# 动态规划的版本



class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        m, n = len(s), len(p)
        dp = [[False] * (n + 1) for _ in range(m + 1)]
        dp[m][n] = True
        for i in range(m, -1, -1):
            for j in range(n-1, -1, -1):
                if p[j] == '*':
                    if i < m:
                        dp[i][j] = dp[i+1][j] or dp[i+1][j+1] or dp[i][j+1]
                    else:
                        dp[i][j] = dp[i][j+1]
                elif i < m and (s[i] == p[j] or p[j] == '?'):
                    dp[i][j] = dp[i+1][j+1]

        return dp[0][0]



# 比之前的那道题更加简单

if __name__ == "__main__":
    cases = [
        ("aa", "a", False),
        ("aa", "*", True),
        ("cb", "?a", False),
        ("adceb", "*a*b", True),
        ("acdcb", "a*c?b?", False),
        ("", "******", True),
    ]
    for s, p, r in cases:
        print(s, p, r)
        t = Solution().isMatch(s, p)
        assert (t == r)

    s = "aaaaaabbaabaaaaabababbabbaababbaabaababaaaaabaaaabaaaabababbbabbbbaabbababbbbababbaaababbbabbbaaaaaaabbaabbbbababbabbaaabababaaaabaaabaaabbbbbabaaabbbaabbbbbbbaabaaababaaaababbbbbaabaaabbabaabbaabbaaaaba"
    p = "*bbb**a*******abb*b**a**bbbbaab*b*aaba*a*b**a*abb*aa****b*bb**abbbb*b**bbbabaa*b**ba**a**ba**b*a*a**aaa"
    print(Solution().isMatch(s, p))
